formats/cbor/commonMain/src/kotlinx/serialization/cbor/Tags.kt

```diff
@@ -0,0 +1,70 @@
+package kotlinx.serialization.cbor
+
+import kotlinx.serialization.cbor.internal.SuppressAnimalSniffer
+import kotlinx.serialization.*
+
+/**
+ * Specifies that a property shall be tagged and serialized as CBOR major type 6: optional semantic tagging
+ * of other major types.
+ *
+ * Example usage:
+ *
+ * ```
+ * @Serializable
+ * data class Data(
+ *     @ValueTags(1337uL)
+ *     @ByteString
+ *     val a: ByteArray, // CBOR major type 6 1337(major type 2: a byte string).
+ *
+ *     @ValueTags(1234567uL)
+ *     val b: ByteArray  // CBOR major type 6 1234567(major type 4: an array of data items).
+ * )
+ * ```
+ *
+ * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+ */
+@SerialInfo
+@Target(AnnotationTarget.PROPERTY)
+@ExperimentalSerializationApi
+public annotation class ValueTags(@OptIn(ExperimentalUnsignedTypes::class) vararg val tags: ULong) {
+    public companion object {
+        public const val DATE_TIME_STANDARD: ULong = 0u;
+        public const val DATE_TIME_EPOCH: ULong = 1u;
+        public const val BIGNUM_POSITIVE: ULong = 2u;
+        public const val BIGNUM_NEGAIVE: ULong = 3u;
+        public const val DECIMAL_FRACTION: ULong = 4u;
+        public const val BIGFLOAT: ULong = 5u;
+        public const val BASE64_URL: ULong = 21u;
+        public const val BASE64: ULong = 22u;
+        public const val BASE16: ULong = 23u;
+        public const val CBOR_ENCODED_DATA: ULong = 24u;
+        public const val URI: ULong = 32u;
+        public const val STRING_BASE64_URL: ULong = 33u;
+        public const val STRING_BASE64: ULong = 34u;
+        public const val REGEX: ULong = 35u;
+        public const val MIME_MESSAGE: ULong = 36u;
+        public const val CBOR_SELF_DESCRIBE: ULong = 55799u;
+    }
+}
+
+/**
+ * Specifies that a key (i.e. a property identifier) shall be tagged and serialized as CBOR major type 6: optional
+ * semantic tagging of other major types.
+ *
+ * Example usage:
+ *
+ * ```
+ * @Serializable
+ * data class Data(
+ *     @KeyTags(34uL)
+ *     val b: Int = -1   // results in the CBOR equivalent of 34("b"): -1
+ * )
+ * ```
+ *
+ * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+ */
+@SerialInfo
+@Target(AnnotationTarget.PROPERTY)
+@ExperimentalSerializationApi
+@SuppressAnimalSniffer // Long.toUnsignedString(long)
+public annotation class KeyTags(@OptIn(ExperimentalUnsignedTypes::class) vararg val tags: ULong)
```

Comment:
I don't feel compiler workarounds should be necessary, but rather the compiler should be fixed.
That aside, while I understand your point of view (Why @CborArray uses @tag, but @KeyTags using ULong) I do feel that having the word "Tag" in the annotation in some form does add to the expressiveness of the source code and conveys intentions in a clear manner. For `@KeyTags` and `@ValueTags` the semantics of the ULongs passed to the annotations are very clear. For `@CBorArray`, plain numbers as arguments do not convey this intent in such a clear manner, imho. Wrapping numbers in  `@Tag` annotations, does. At the same time `@KeyTags(Tag(1337uL))` is just needlessly verbose compared to `@KeyTags(1337uL)` and brings no added value.

I will admit that I am not 100% happy with this solution, but it was the most expressive and concise concept I could come up with.

Reply:
I'm sure the compiler will be fixed eventually (in 2.1 or so, judging by the ticket activity). The argument that 'Tag' is already in the `@KeyTags` name also sounds reasonable. Have you considered also having `@CborArray + @CborArrayTags`, or is it too verbose?

Btw, don't we need tags for regular objects (maps)? Maybe we can have `@ObjectTags`?

Reply:
I think this is what being asked by @Isolus :
> I see examples to tag classes when I want them to be arrays with @CborArray(Tag()). Can I tag classes when they are maps somehow?

But I don't get it why you've answered that this won't work. As far as I understood, `@CborArray(Tag(..))` appends tag to a whole array structure — as if `@ValueTags` was specified on a property of this type. So why we shouldn't be able to do the same with regular objects?

Reply:
Now I understand. The question by @Isolus now also makes much more sense and so does the idea of introducing tagging to objects. `@ObjectTags` is added in aeb83cf3fb9bd5a9684a172567a3bcc0b12c5567.
I have also unified it so that `@CborArray` not doesn't take tags anymore, but instead `@ObjectTags` can be used independently of the `@CborArray` annotation.